### PR TITLE
Send post button changes

### PR DESCRIPTION
**Pull request by harshilsharma63**
#### Summary
* Split the send post button into two parts, one for sending post and another one for scheduling post.
* Added pre-defined, dynamic options for a scheduled time based on user's day of the week.
* Added functionality to create a scheduled post one an option is selected, and show error message if it fails.

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-60119

#### Screenshots
Screenshot 2024-08-21 at 3 20 39 PM

#### Release Note
```release-note
None
```

### CODE REVIEW

File: webapp/channels/src/actions/views/create_comment.tsx [hunk 1]
```diff
@@ -103,6 +105,25 @@ export function submitPost(channelId: string, rootId: string, draft: PostDraft,
 
         post = hookResult.data;
 
+        if (schedulingInfo) {
```
Comment (hmhealey) [inline webapp/channels/src/actions/views/create_comment.tsx:108]: Perhaps instead of having to pass `schedulingInfo` all the way through `onSubmit` and `submitPost`, we could instead reorganize this so that all of the checking for channel mentions and running hooks are done by one or more helper functions that we can call from both `submitPost` and some new `schedulePost`? Moving those into helper functions that can be called by two different actions is some extra work, but it'll help make it easier to understand what all happens when you click the submit buttons.

I want to avoid the type of issue we run into with the `Post` component where it's full of complex if statements because we went in the direction of a single `Post` component that can behave a variety of different ways due to various combinations of parameters instead of making multiple components like `RhsPost`, `CenterChannelPost`, and `SearchPost` that share children. In both that case and this one, I think it's better to have two single purpose pieces of code that use the same helpers in the same order rather than to add conditionals like that since you get code with fewer pieces of branching logic that need to be understood.

Thoughts? I'm open to discussing this as well since I'm aware that I'm also asking you to help clean up a bunch of existing code

File: webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx [hunk 2]
```diff
@@ -43,32 +39,74 @@ const SendButtonContainer = styled.button`
...
+                id='send_post_now_tooltip'
+                title={formatMessage({id: 'create_post_button.option.send_now', defaultMessage: 'Send Now'})}
+                shortcut={sendNowKeyboardShortcutDescriptor}
+                disabled={disabled}
+            >
+                <SendButtonContainer
+                    className={classNames('SendMessageButton', {disabled})}
+                    data-testid='SendMessageButton'
+                    tabIndex={0}
+                    aria-label={formatMessage({
```
Comment (hmhealey) [inline webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx:87]: This should probably be `Send Now` as well

File: webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx [hunk 3]
```diff
@@ -43,32 +39,74 @@ const SendButtonContainer = styled.button`
...
+                        id: 'create_post.send_message',
+                        defaultMessage: 'Send a message',
+                    })}
+                    disabled={disabled}
+                    onClick={sendMessage}
+                >
+                    <SendIcon
+                        size={18}
+                        color='currentColor'
+                        aria-label={formatMessage({
```
Comment (hmhealey) [inline webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx:97]: Could you also remove this label? It's not read out by Chrome's screen reader, but it'd probably be good to prevent it from reading out "Send now create a post" one after another if it did read it

File: webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx [hunk 4]
```diff
@@ -1,39 +1,35 @@
...
+import type {SchedulingInfo} from '@mattermost/types/schedule_post';
+
+import {isSendOnCtrlEnter} from 'selectors/preferences';
+
+import {SendPostOptions} from 'components/advanced_text_editor/send_button/send_post_options';
+import WithTooltip from 'components/with_tooltip';
+import type {ShortcutDefinition} from 'components/with_tooltip/shortcut';
+import {ShortcutKeys} from 'components/with_tooltip/shortcut';
+
+import './style.scss';
```
Comment (hmhealey) [inline webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx:20]: Could we avoid styling this with both Styled Components and CSS? I know we're not consistent with the two, but I think it'd be better to not mix the two on a single element

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        const timeComponent = (
+            <Timestamp
+                value={unixTimestamp}
+                useDate={false}
+            />
+        );
+
+        const tomorrow9amTime = new Date();
+        tomorrow9amTime.setHours(9, 0, 0, 0);
+        tomorrow9amTime.setDate(today.getDate() + 1);
```
Comment (hmhealey) [inline webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx:46]: I think we might want to do the `setDate` first because of daylight savings time (the programmer's worst enemy)

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+        const optionTomorrow = (
+            <Menu.Item
+                onClick={(e) => handleOnSelect(e, tomorrow9amTime.getTime())}
+                labels={
+                    <FormattedMessage
+                        id='create_post_button.option.schedule_message.options.tomorrow'
+                        defaultMessage='Tomorrow at {9amTime}'
+                        values={{'9amTime': timeComponent}}
```
Comment (hmhealey) [inline webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx:55]: Are you sure we need to use the timestamp here? I would've expected that this would just be part of the translation string since the translator would know how to translate it. It seems weird to use the `Timestamp` component here when it includes so much logic in it which isn't otherwise needed here

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+                }
+            />
+        );
+
+        const options: React.ReactElement[] = [];
+
+        switch (today.getDay()) {
+        // Sunday
+        case 0:
+            options.push(optionTomorrow);
```
Comment (hmhealey) [inline webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx:100]: Nitpicking, but it might be nicer to right this as
```js
let options;

switch (today.getDay()) {
case 0:
    options = [optionTomorrow];
    break;
    ...
}

return options;
```

I think that should still type correctly, and I think using the array literals is a bit nicer than pushing onto the array.

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss [hunk 8]
```diff
@@ -0,0 +1,9 @@
+ul#dropdown_send_post_options {
+    li[role="menuitem"][aria-disabled="true"] {
```
Comment (hmhealey) [inline webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss:2]: Could these styles be applied directly to the menu component instead of this being a special case? I don't know how if it currently supports disabled elements, but I don't see why it couldn't

File: webapp/channels/src/components/advanced_text_editor/use_submit.tsx [hunk 9]
```diff
@@ -185,27 +190,27 @@ const useSubmit = (
...
+    const showNotifyAllModal = useCallback((mentions: string[], channelTimezoneCount: number, memberNotifyCount: number, schedulingInfo?: SchedulingInfo) => {
         dispatch(openModal({
             modalId: ModalIdentifiers.NOTIFY_CONFIRM_MODAL,
             dialogType: NotifyConfirmModal,
             dialogProps: {
                 mentions,
                 channelTimezoneCount,
                 memberNotifyCount,
-                onConfirm: () => doSubmit(),
+                onConfirm: () => doSubmit(undefined, undefined, schedulingInfo),
```
Comment (hmhealey) [inline webapp/channels/src/components/advanced_text_editor/use_submit.tsx:208]: Instead of having `showNotifyAllModal` depend on `doSubmit`, I think it might be better for it to take a callback that tells it what to do like `prioritySubmitCheck` does.

Related to my comment about having two actions for `submitPost` and `schedulePost` though, the fact that we're having to pass around and consider `schedulingInfo` here feels like a red flag, especially since we have to pass `undefined` through as the first two arguments

File: webapp/channels/src/components/advanced_text_editor/use_submit.tsx [hunk 10]
```diff
@@ -238,70 +243,72 @@ const useSubmit = (
             channelTimezoneCount = data ? data.length : 0;
         }
 
-        if (prioritySubmitCheck(doSubmit)) {
+        if (prioritySubmitCheck(() => doSubmit(undefined, undefined, schedulingInfo))) {
```
Comment (hmhealey) [inline webapp/channels/src/components/advanced_text_editor/use_submit.tsx:246]: This is an existing bug which I'll file a ticket for, but I just noticed that any of these checks that are async and use a callback skip the following checks so, in this case, if you have a priority message, you'll never be prompted about at-mentioning too many people

File: webapp/channels/src/components/advanced_text_editor/use_submit.tsx [hunk 11]
```diff
@@ -238,70 +243,72 @@ const useSubmit = (
...
 
-        const status = getStatusFromSlashCommand(submittingDraft.message);
-        if (userIsOutOfOffice && status) {
-            const resetStatusModalData = {
-                modalId: ModalIdentifiers.RESET_STATUS,
-                dialogType: ResetStatusModal,
-                dialogProps: {newStatus: status},
-            };
+        if (!schedulingInfo) {
+            const status = getStatusFromSlashCommand(submittingDraft.message);
```
Comment (hmhealey) [inline webapp/channels/src/components/advanced_text_editor/use_submit.tsx:258]: This is more existing stuff, but I don't think I realized that this hook is the one responsible for a bunch of these clientside slash commands. I would've expected `submitCommand` to do that

File: webapp/channels/src/components/menu/menu.tsx [hunk 12]
```diff
@@ -65,14 +65,17 @@ type MenuProps = {
 const defaultAnchorOrigin = {vertical: 'bottom', horizontal: 'left'};
 const defaultTransformOrigin = {vertical: 'top', horizontal: 'left'};
 
-type VerticalOrigin = 'top' | 'center' | 'bottom';
-type HorizontalOrigin = 'left' | 'center' | 'right';
+type VerticalOrigin = 'top' | 'center' | 'bottom' | number;
+type HorizontalOrigin = 'left' | 'center' | 'right' | number;
```
Comment (hmhealey) [inline webapp/channels/src/components/menu/menu.tsx:69]: Are these still needed? I don't see you passing any numbers for these elsewhere

File: webapp/channels/src/components/menu/menu.tsx [hunk 13]
```diff
@@ -303,7 +310,7 @@ function MenuModal(props: MenuModalProps) {
         <CompassDesignProvider theme={theme}>
             <GenericModal
                 id={props.menuId}
-                className='menuModal'
+                className='menuModal LOOOOOOOOOOOOOOOOOOOOOOOOOOOOOL'
```
Comment (hmhealey) [inline webapp/channels/src/components/menu/menu.tsx:313]: 😆

File: webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts [hunk 14]
```diff
@@ -298,7 +299,7 @@ export function createPost(post: Post, files: any[] = [], afterSubmit?: (respons
                 dispatch(batchActions(actions, 'BATCH_CREATE_POST_FAILED'));
                 return {error};
             }
-        }());
+        }()).then(() => {}));
```
Comment (hmhealey) [inline webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts:302]: What's this change for?

File: webapp/channels/src/selectors/preferences.ts [hunk 15]
```diff
@@ -15,3 +15,21 @@ export const arePreviewsCollapsed = (state: GlobalState) => {
         Preferences.COLLAPSE_DISPLAY_DEFAULT !== 'false',
     );
 };
+
+export const isSendOnCtrlEnter = (state: GlobalState) => {
```
Comment (hmhealey) [inline webapp/channels/src/selectors/preferences.ts:19]: Are we getting these values using `getBoolPreference` anywhere else? If so, could you either change those now or file a ticket to do that later?

We might also need to move these into mattermost-redux to do that

File: webapp/platform/client/src/client4.ts [hunk 16]
```diff
@@ -4314,6 +4315,14 @@ export default class Client4 {
             {method: 'post', body: JSON.stringify(body)},
         );
     };
+
+    // Schedule Post methods
+    createSchedulePost = (schedulePost: ScheduledPost, connectionId: string) => {
```
Comment (hmhealey) [inline webapp/platform/client/src/client4.ts:4320]: Why are we passing the `connectionId` here?

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: server/channels/app/scheduled_post.go [hunk 17]
```diff
@@ -11,6 +11,11 @@ import (
 )
 
 func (a *App) SaveScheduledPost(rctx request.CTX, scheduledPost *model.ScheduledPost) (*model.ScheduledPost, *model.AppError) {
+	scheduledPost.PreSave()
+	if validationErr := scheduledPost.BaseIsValid(); validationErr != nil {
```
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply server/channels/app/scheduled_post.go:15]: We check it in store level, not app level. If you see the IsValid function, it accepts the allowed max length of message, which is available only in the SQL store layer because its computed by querying the database for max allowed length of message column in the database. So, on app layer I'm validating everything except the message length, and message length is being checked on store layer along with all other checks (yes its duplicate check but its fine).

Previous all checks were only at store layer, causing a failure to communicate the validation error in API responses, so I moved all validations minus message length here. This is an improvement how we validate drafts and posts.

File: webapp/channels/src/actions/post_actions.ts [hunk 18]
```diff
@@ -106,14 +110,19 @@ export function unflagPost(postId: string): ActionFuncAsync {
     };
 }
 
-export function createPost(post: Post, files: FileInfo[], afterSubmit?: (response: SubmitPostReturnType) => void): ActionFuncAsync<PostActions.CreatePostReturnType, GlobalState> {
+function parseMessageEmojis(message: string, dispatch: Dispatch) {
```
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply webapp/channels/src/actions/post_actions.ts:113]: Done.

File: webapp/channels/src/actions/post_actions.ts [hunk 19]
```diff
@@ -127,6 +136,22 @@ export function createPost(post: Post, files: FileInfo[], afterSubmit?: (respons
     };
 }
 
+export function createSchedulePostFromDraft(scheduledPost: ScheduledPost): ActionFuncAsync<PostActions.CreatePostReturnType, GlobalState> {
```
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply webapp/channels/src/actions/post_actions.ts:139]: Its doing so because I need process the error occurring in creating the scheduled posts and for type correctness. If I return something else, I need to update the type in a lot of places, basically in all the functions the entire chain of post button till the API call and the same in post/edit comment. Doing so wasn't providing any benefit.

File: webapp/channels/src/actions/schedule_message.ts [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+import {Client4} from 'mattermost-redux/client';
+import type {DispatchFunc, GetStateFunc} from 'mattermost-redux/types/actions';
+
+import {getConnectionId} from 'selectors/general';
+
+import type {GlobalState} from 'types/store';
+
+export function createSchedulePost(schedulePost: ScheduledPost) {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        const state = getState() as GlobalState;
```
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply webapp/channels/src/actions/schedule_message.ts:15]: Done.

File: webapp/channels/src/actions/schedule_message.ts [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+            // TODO: dispatch action to store created schedule
+            // post in store to display it.
+        } catch (error) {
+            return {
+                created: false,
+                error,
+            };
+        }
+
+        return {created: true};
```
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply webapp/channels/src/actions/schedule_message.ts:30]: Done, returning the data now.

File: webapp/channels/src/actions/views/create_comment.tsx (hunk 1 above)
Comment (harshilsharma63) [reply webapp/channels/src/actions/views/create_comment.tsx:108]: I've gone through this discussion with Daniel and what we have now is the cleanest version possible. I initially did start by implementing a different hook for processing scheduled post, but if as you can see, all we need to do differently is to skip just 3-4 steps in the whole chain of function calls. Splitting this into helper functions and having two separate hooks uses those helpers sounds good, but I am sure it will cause the issue of someone making a change in send post logic without testing and making the same change in schedule post logic. Current code avoid that issue.

If you go through the whole function chain, you'll realise its not a lot of if checks and there is no mixing of logic.

File: webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx (hunk 2 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx:87]: `Send a message` is more descriptive of the button's functionality than `Send Now`. In case of screen reader, send a message is a better description.

Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx:87]: Oh, I added the missing aria label on schedule button.

File: webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx (hunk 3 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx:97]: Done.

File: webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx (hunk 4 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx:20]: Okay, IaI've removed the styled component now and we're all SCSS now 🙌

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx (hunk 5 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx:46]: Ah thanks! You saved a couple of hours for future me 🙏

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx (hunk 6 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx:55]: We need timestamp here as we need to show the time as per user's configured timezone and based on the user's preference of 12 or 24 hour time format.

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx (hunk 7 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx:100]: Done.

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss (hunk 8 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss:2]: By default it greys out, and doesn't bold disabled elements, both things which we wanna customise as per designs. Styling via CSS selectors is significantly easier, more readable and much more powerful than passing styles in JSX code to the component.

File: webapp/channels/src/components/advanced_text_editor/use_submit.tsx (hunk 9 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/use_submit.tsx:208]: Isn't this already taking a callback, via dialogProps? Callbacks to NotifyConfirmModal and prioritySubmitCheck are exactly the same. Or am I missing something.

File: webapp/channels/src/components/advanced_text_editor/use_submit.tsx (hunk 10 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/use_submit.tsx:246]: Its sounds wrong but it works correctly. Priority with bulk mention does indeed trigger the alert about at mentioning everyone because `prioritySubmitCheck` returns false unless you have set persistent notification, and if you do set persistent notification, you aren't allowed to bulk mention anywhere, so it works correctly. So everything is working correctly.

File: webapp/channels/src/components/advanced_text_editor/use_submit.tsx (hunk 11 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/use_submit.tsx:258]: Well, it it what it is, and it works and has been working for ages.

File: webapp/channels/src/components/menu/menu.tsx (hunk 12 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/menu/menu.tsx:69]: No, they aren't needed. Reverted.

File: webapp/channels/src/components/menu/menu.tsx (hunk 13 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/menu/menu.tsx:313]: Sometimes the most trivial methods of debugging are the most effective 😀

Comment (harshilsharma63) [reply webapp/channels/src/components/menu/menu.tsx:313]: Removed.

File: webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts (hunk 14 above)
Comment (harshilsharma63) [reply webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts:302]: This isn't needed. It was my just my IDE complaining and I auto-fixed it. Reverted it now.

File: webapp/channels/src/selectors/preferences.ts (hunk 15 above)
Comment (harshilsharma63) [reply webapp/channels/src/selectors/preferences.ts:19]: Currently use use the generic `getPreference` selector from redux store to get some of these values, and for some newer settings we have these functional selectors outside the `mattermost-redux` package.

Isn't the idea to get rid of the mattermost-redux package and add new redux stuff in `channels/src/selectors/`?

File: webapp/platform/client/src/client4.ts (hunk 16 above)
Comment (harshilsharma63) [reply webapp/platform/client/src/client4.ts:4320]: The API will use it to filter out the current client's WS connection when sending the WS event. This is to avoid duplicate data in store, as being done in some other APIs as well.

Comment (hmhealey) [reply webapp/platform/client/src/client4.ts:4320]: That's what I was hoping since I've wanted this for so many other APIs 😀

Instead of passing it as an argument to this one method, what if we saved the `connectionId` on the `Client4` instance and then have that header automatically attached to all requests? It shouldn't hurt them, but it'd set us up to use this elsewhere pretty nicely while keeping `createSchedulePost` simpler

Comment (hmhealey) [reply webapp/platform/client/src/client4.ts:4320]: Also, I just realized the name of this. It's got a small typo, so could you rename it to `createScheduledPost`

File: webapp/channels/src/selectors/preferences.ts (hunk 15 above)
Comment (hmhealey) [reply webapp/channels/src/selectors/preferences.ts:19]: I mostly dislike the way that plugins use mattermost-redux, but I'm undecided on the package itself in the web app. The only reason I mentioned it here is in case something else in mattermost-redux needs this because it can't import anything from outside of itself

File: server/channels/app/scheduled_post.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply server/channels/app/scheduled_post.go:15]: Ah, I see. In that case, perhaps instead of splitting up the method, we could just pass some arbitrarily large number here and comment that it's double checked later? 0/5

File: webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx (hunk 2 above)
Comment (hmhealey) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx:74]: I can sort of understand that, but I would think that that should match the tooltip though since the tooltip has the same general purpose as the label. I'm also thinking it's like how the `title` attribute adds a system tooltip and is also read by a screen reader

File: webapp/channels/src/actions/views/create_comment.tsx (hunk 1 above)
Comment (hmhealey) [reply webapp/channels/src/actions/views/create_comment.tsx:108]: That makes sense to me. I do think reducing the complexity of that flow might be worth the tradeoff of making it so that people need to add things in two places, but since most of that would require reorganizing a lot of existing code, that's definitely outside the scope of what we want to do here

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx (hunk 6 above)
Comment (hmhealey) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/index.tsx:56]: Shouldn't it already be 9am in their time zone because it comes from a `Date` object? That's a good point about the 12 or 24 hour time format though.

It does seem a bit funny that we're passing the same 9am tomorrow `Timestamp` component into the `FormattedMessage` for "Monday at 9:00", but that shouldn't cause any problems

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss (hunk 8 above)
Comment (hmhealey) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss:2]: Sorry, I meant to say "could this be applied to everywhere this Menu component is used instead of just this one?"

I can't think of anywhere else that we disable menu items, but I think it'd be good to ensure that that's consistently applied in the future. We could double check with UX on that, but I don't expect them to want this menu to be different from others

File: webapp/channels/src/components/advanced_text_editor/use_submit.tsx (hunk 9 above)
Comment (hmhealey) [reply webapp/channels/src/components/advanced_text_editor/use_submit.tsx:208]: Sorry, I meant that we pass an `onConfirm` callback into `showNotifyAllModal` instead of having it call `doSubmit` directly like the one that `prioritySubmitCheck` takes. It still risks skipping followup logic like I mentioned below, but it saves us from having this callback depend on `doSubmit` and needing to take `schedulingInfo`. 0/5 though on actually changing it

File: server/channels/app/scheduled_post.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply server/channels/app/scheduled_post.go:15]: I don't think it provides any benefit and the check will obviously wrong, so even if isValid returned true in app layer, you still aren't sure if the entire object is valid, kinda faking the meaning of Isvalid. I'd prefer to not do it.

File: webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx (hunk 2 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_button.tsx:74]: I see, that makes sense. Done.

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss (hunk 8 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss:2]: We don't want to do it. You do want disabled menu items to appear disabled, right? This here in this component is the exception where I'm using a disabled menu item as a way to display a text with same font size, same height and everything. Setting this style globally will make all disabled and enabled elements look alike.

File: webapp/channels/src/components/advanced_text_editor/use_submit.tsx (hunk 9 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/use_submit.tsx:208]: Ah I understand now. Done!

File: webapp/channels/src/components/advanced_text_editor/use_submit.tsx (hunk 10 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/use_submit.tsx:246]: Yeah, thats true.

File: webapp/channels/src/selectors/preferences.ts (hunk 15 above)
Comment (harshilsharma63) [reply webapp/channels/src/selectors/preferences.ts:19]: No, nothing in the redux package uses this.

File: webapp/platform/client/src/client4.ts (hunk 16 above)
Comment (harshilsharma63) [reply webapp/platform/client/src/client4.ts:4320]: Thats a nice idea, its something we can do in the fixathon.

File: webapp/package-lock.json [hunk 22]
```diff
@@ -23719,7 +23719,7 @@
         "typescript": "^5.0.0"
       },
       "peerDependencies": {
-        "@mattermost/types": "9.3.0",
+        "@mattermost/types": "^9.3.0 || ^10.0.0",
```
Comment (M-ZubairAhmed) [inline webapp/package-lock.json:23722]: lets revert this change

File: webapp/channels/src/components/menu/menu.tsx [hunk 23]
```diff
@@ -205,7 +212,7 @@ export function Menu(props: Props) {
                     id={props.menuButtonTooltip.id}
                     title={props.menuButtonTooltip.text}
                     placement={props?.menuButtonTooltip?.placement ?? 'top'}
-                    disabled={isMenuOpen}
+                    disabled={props.hideTooltipWhenDisabled ? props.menuButton.disabled || isMenuOpen : isMenuOpen}
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/menu/menu.tsx:215]:
```suggestion
                    disabled={props.hideTooltipWhenDisabled ? props.menuButton.disabled  : isMenuOpen}
```
shouldnt this be just this?

File: webapp/channels/src/components/menu/menu.tsx [hunk 24]
```diff
@@ -94,6 +98,9 @@ interface Props {
  * </Menu.Item>
  */
 export function Menu(props: Props) {
+    // Use MUI Anchor Playgroup to try various anchorOrigin
+    // and transformOrigin values - https://mui.com/material-ui/react-popover/#anchor-playground
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/menu/menu.tsx:102]: repeating comment

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss (hunk 8 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss:2]: Why do you want to use disabled styled here?

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/channels/src/actions/schedule_message.ts [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+export function createSchedulePost(schedulePost: ScheduledPost): ActionFuncAsync<{data?: ScheduledPost; error?: string}, GlobalState> {
+    return async (dispatch, getState) => {
+        const state = getState();
+        const connectionId = getConnectionId(state);
+
+        try {
+            const createdPost = await Client4.createScheduledPost(schedulePost, connectionId);
+            return {data: createdPost};
+
+            // TODO: dispatch action to store created schedule
```
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply webapp/channels/src/actions/schedule_message.ts:22]: Thats to be done later. Not part of this PR. I keep adding TODO comments for stuff I know I have to do later, but not part of this PR.

File: webapp/platform/types/src/schedule_post.ts [hunk 26]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import type {Draft} from './drafts';
+
+export type SchedulingInfo = {
+    scheduled_at: number;
```
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply webapp/platform/types/src/schedule_post.ts:7]: I see snake case everywhere in all the types.

[the remark this answers is not in the document]
Comment (harshilsharma63) [reply webapp/platform/types/src/schedule_post.ts:7]: Was there a decision somewhere to migrate to camel case? I still see all new types using camel case, so continuing with it.

File: webapp/package-lock.json (hunk 22 above)
Comment (harshilsharma63) [reply webapp/package-lock.json:23722]: My bad, done. reverted.

File: webapp/channels/src/components/menu/menu.tsx (hunk 23 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/menu/menu.tsx:215]: No, this will cause the button tooltip to continue showing up unless its disabled by props. Hiding the tooltip automatically on opening the manu is an existing behaviour which would be broken with this suggestion.

File: webapp/channels/src/components/menu/menu.tsx (hunk 24 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/menu/menu.tsx:102]: Done.

File: webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss (hunk 8 above)
Comment (harshilsharma63) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss:2]: Because I'm altering a disabled menu item. I want to display a text. I can use a div or span but will have to manage its all styling like width, height, font etc to match that of other menu items. So, I use a manu item and disabled it so it appears as a text and is disabled so you can't select it.

Comment (hmhealey) [reply webapp/channels/src/components/advanced_text_editor/send_button/send_post_options/style.scss:2]: Oh, I see. I didn't realize that that's why the menu item was disabled.

That seems reasonable to me then. We'll need a generic `Menu.Heading` component eventually for some other menus in the app, but I don't think we need to add that right now. The disabled menu item is a bit funny looking, but that's totally fine.

I also found that you can pass a `subheader` into the `MuiMenu`'s `MenuListProps`, but it also needs custom styling unless you use their `ListSubheader` component which defaults to the tiny text that Android menu headings have, so it has the same issue as using an arbitrary disabled div

**Review by hmhealey — APPROVED** [review]

File: webapp/platform/types/src/schedule_post.ts (hunk 26 above)
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply webapp/platform/types/src/schedule_post.ts:7]: In this case this is being sent (and in future be received from) server, so snake case is fine.

**Review by M-ZubairAhmed — APPROVED** [review]
